lib/elixir/lib/kernel/typespec.ex

```diff
@@ -92,6 +92,12 @@ defmodule Kernel.Typespec do
     end
   end
 
+  defmacro defoptional_callbacks(callbacks) do
+    quote do
+      Kernel.Typespec.defspec(:optional_callbacks, unquote(Macro.escape(callbacks, unquote: true)), __ENV__)
```

Comment:
You don't need to escape here. `callbacks` here are just values and not ASTs as in spec, type and friends. It is very likely this could be implemented like:

    Kernel.Typespec.defoptional_callbacks(unquote(callbacks), __ENV__)

and then:

    def defoptional_callbacks(callbacks, env) do
      Module.put_attribute(env.module, [{env.line, callbacks} | Module.get_attribute(env.module, :optional_callbacks) || []])
    end

If you mark it as a accumulated attribute [here](https://github.com/elixir-lang/elixir/blob/master/lib/elixir/src/elixir_module.erl#L161), the implementation can even be just: 

    def defoptional_callbacks(callbacks, env) do
      Module.put_attribute(env.module, {env.line, callbacks})
    end